src/write/abbrev.rs

```diff
@@ -0,0 +1,222 @@
+use indexmap::IndexSet;
+use std::ops::{Deref, DerefMut};
+use vec::Vec;
+
+use common::DebugAbbrevOffset;
+use constants;
+use write::{Result, Section, SectionId, Writer};
+
+/// A table of abbreviations that will be stored in a `.debug_abbrev` section.
+// Requirements:
+// - values are `Abbreviation`
+// - insertion returns an abbreviation code for use in writing a DIE
+// - inserting a duplicate returns the code of the existing value
+#[derive(Debug, Default)]
+pub(crate) struct AbbreviationTable {
+    abbrevs: IndexSet<Abbreviation>,
+}
+
+impl AbbreviationTable {
+    /// Add an abbreviation to the table and return its code.
+    pub fn add(&mut self, abbrev: Abbreviation) -> u64 {
+        let (code, _) = self.abbrevs.insert_full(abbrev);
+        // Code must be non-zero
+        (code + 1) as u64
+    }
+
+    /// Write the abbreviation table to the `.debug_abbrev` section.
+    pub fn write<W: Writer>(&self, w: &mut DebugAbbrev<W>) -> Result<()> {
+        for (code, abbrev) in self.abbrevs.iter().enumerate() {
+            w.write_uleb128((code + 1) as u64)?;
+            abbrev.write(w)?;
+        }
+        // Null abbreviation code
+        w.write_u8(0)
+    }
+}
+
+/// An abbreviation describes the shape of a `DebuggingInformationEntry`'s type:
+/// its tag type, whether it has children, and its set of attributes.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub(crate) struct Abbreviation {
+    tag: constants::DwTag,
+    has_children: bool,
+    attributes: Vec<AttributeSpecification>,
+}
+
+impl Abbreviation {
+    /// Construct a new `Abbreviation`.
+    #[inline]
+    pub fn new(
+        tag: constants::DwTag,
+        has_children: bool,
+        attributes: Vec<AttributeSpecification>,
+    ) -> Abbreviation {
+        Abbreviation {
+            tag,
+            has_children,
+            attributes,
+        }
+    }
+
+    /// Write the abbreviation to the `.debug_abbrev` section.
+    pub fn write<W: Writer>(&self, w: &mut DebugAbbrev<W>) -> Result<()> {
+        w.write_uleb128(self.tag.0)?;
+        w.write_u8(if self.has_children {
+            constants::DW_CHILDREN_yes.0
```

Comment:
It seems our `dw!` macro for constants could also generate a thin little `write` method. Although, I suppose most of the time we want to write LEB128s, so maybe it wouldn't be super useful and this location would still be the exception.

Reply:
I'm not sure about the usefulness:
- they would be simple one line functions
- many of them could be written in multiple forms that the caller would need to decide, since attribute values could potentially use either LEB128 or a fixed size, and this needs to match the form, so it's better to have this code closer to the form decision.